update what things are encrypted (#19)
suhasHere/moq-secure-objects: draft-jennings-moq-secure-objects.md

1. ObjectNames are globally unique in a given relay network.

> 2. The data inside an object (and it's size) can never change after the

Reviewer: ```suggestion
2. The data inside an object (and its size) can never change after the
```

# Secure Objects

{{Section 7.1.1 of I-D.ietf-moq-transport}} defines fields of a canonical
MOQT Object. The protection scheme defined in this draft encrypts the

> `Object Payload` and the fields `Track Alias`, `Group ID`, `Object ID`,
> `Object Payload` are authenticated, regardless of the on-the-wire encoding
> of the objects over QUIC Datagrams or QUIC streams.

Reviewer: ```suggestion
`Object Payload` and authenticates the `Track Alias`, `Group ID`, `Object ID`,
and `Object Payload` fields, regardless of the on-the-wire encoding
of the objects over QUIC Datagrams or QUIC streams.
```

> TODO: Move authentication tag to Object Header Extension.

Reviewer: ```suggestion
```

No, this should not be done.  According to the AEAD interface in [RFC 5116](https://datatracker.ietf.org/doc/html/rfc5116#section-2.1), there is no separate authentication tag, just a unitary ciphertext.

## Setup Assumptions

> We assume that the application assigns each track a set of (KID, base_key)

Reviewer: ```suggestion
We assume that the application assigns each track a set of (KID, `base_key`)
```

This change is incorrect.

}
~~~

> TODO: Make Key ID as Object Header Extension

Reviewer: Don't add a TODO, either fix it or don't, preferably the former.

## Metadata Authentication

> The KID, fulltrackname, Group ID, and Object ID for a given object are authenticated as

Reviewer: ```suggestion
The KID, FullTrackName, Group ID, and Object ID for a given object are authenticated as
```

"fulltrackname" is not a word.

part of secure object encryption.  This ensures, for example, that encrypted
objects cannot be replayed across tracks.

When protecting or unprotecting a secure object, an endpoint encodes the key ID,

> Group ID, Object ID, and fulltrackname in the following data structure, for

Reviewer: ```suggestion
Group ID, Object ID, and FullTrackName in the following data structure, for
```

~~~ pseudocode
def decrypt(full_track_name, object):

>     # Parse the secure object payload to obtain kid

Reviewer: ```suggestion
    # Parse the secure object payload to obtain key ID and ciphertext
```

    (kid, kid_byte_len) = parse_varint(object.payload)
    ciphertext = object.payload[kid_byte_len:]

>     # Identify the appropriate encryption context mapped to the track alias
>     # and the kid

Reviewer: ```suggestion
    # Identify the appropriate encryption context for the full track name
    # and the key ID
```

The SFrame specification lists several things that an application needs to
account for in order to use SFrame securely, which are all accounted for here:

1. **Header value uniqueness:** Uniqueness of CTR values follows from the

>    uniqueness of MOQT (group ID, object ID). We only use one KID value, but

Reviewer: ```suggestion
   uniqueness of MOQT (group ID, object ID) pairs. We only use one KID value, but
```

The pair is the thing that is unique.

   the uniqueness constraints on object IDs and objects, and because the group
   ID and object ID are cryptographically bound to the secure object payload.

> 4. **Metadata:** The content of the metadata input to Secure Object operations is

Reviewer: ```suggestion
4. **Metadata:** The analogue of the SFrame metadata input is
```
